variable/bin_detail.cpp

```diff
@@ -0,0 +1,87 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// Copyright (c) 2021 Scipp contributors (https://github.com/scipp)
+/// @file
+/// @author Simon Heybrock
+#include "scipp/core/element/bin_detail.h"
+#include "scipp/core/element/util.h" // fill_zeros
+#include "scipp/core/subbin_sizes.h"
+
+#include "scipp/variable/bin_detail.h"
+#include "scipp/variable/bin_util.h"
+#include "scipp/variable/subspan_view.h"
+#include "scipp/variable/transform.h"
+#include "scipp/variable/util.h"
+
+namespace scipp::variable::bin_detail {
+
+/// Index of the bin (given by `edges`) containing a coord value.
+///
+/// 0 if the coord is less than the first edge, nbin-1 if greater or equal last
+/// edge. Assumes both `edges` and `coord` are sorted.
+Variable begin_edge(const VariableConstView &coord,
+                    const VariableConstView &edges) {
+  auto indices = makeVariable<scipp::index>(coord.dims());
+  const auto dim = edges.dims().inner();
+  Variable bin(indices.slice({dim, 0}));
```

Comment:
What if dim is empty? (Same in `end_edge`)

Reply:
Nice catch! It did not actually get here, but that was due to inconsistent behavior of slicing `DataArray` or `Dataset` with length=0: Bin-edges were preserved as length=1 (since edge coords are always longer by one), but in this special case we actually need to set them to length=0 as well. See updates.